### PR TITLE
Bug 1633053 - Fix/enable ESLint react/prop-types

### PR DESCRIPTION
**Pull request by bolah2009**
## Description

This PR removes the rule `'react/prop-types': 'off'` from `.eslintrc.js` file and fix the prop-types errors by adding missing prop-types validations.

## Proposed Solution

- Prop-types validations are added using [`PropTypes`](https://reactjs.org/docs/typechecking-with-proptypes.html) fo all props without prop-types validation.

## Testing

The react developer tool is used to locate and inspect the components' `prop` and the usage of the `prop` in the component is also observed, before changing the `PropTypes`. Also, there's also a console warning when a wrong prop-type is used to confirm the validity of the `PropTypes` been used. 

I also ran `yarn test` to confirm that none of the existing tests was affected by these changes.

## Resources used
- [Typechecking With PropTypes](https://reactjs.org/docs/typechecking-with-proptypes.html)
- [Forbid certain propTypes (react/forbid-prop-types)](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/forbid-prop-types.md)

-----
I have some questions, I will leave some in-line comments in the diff for this.

### CODE REVIEW

File: ui/perfherder/graphs/TableView.jsx [hunk 1]
```diff
@@ -101,44 +161,11 @@ const TableView = ({
...
-          );
-        }
-        return cellElem;
-      },
-      getProps: (state, rowInfo) => setHighlightedRow(rowInfo),
+      Cell: TestCell,
+      getProps: (state, rowInfo) => ({
+        ...setHighlightedRow(rowInfo),
+        dataKey,
+      }),
```
Comment (bolah2009) [inline ui/perfherder/graphs/TableView.jsx:168]: I couldn't define the PropTypes for `props` of the anonymous arrow function, so I had to extract it to a separate named function following the suggestion from https://github.com/yannickcr/eslint-plugin-react/issues/2404#issuecomment-529232335

File: ui/perfherder/tests/TestsTableControls.jsx [hunk 2]
```diff
@@ -79,6 +79,7 @@ export default class TestsTableControls extends React.Component {
 }
 
 TestsTableControls.propTypes = {
+  validated: PropTypes.shape({}),
```
Comment (bolah2009) [inline ui/perfherder/tests/TestsTableControls.jsx:82]: I have a question about the use of `validated` here.

I was trying to check for the data type in the react dev tools but noticed `validated` is not passes as a prop to the `TestsTableControls` component when used in `TestsView` component (See below)

https://github.com/mozilla/treeherder/blob/d6ba99312cb954bf07db3c0f6f769f72ca7cdfe0/ui/perfherder/tests/TestsView.jsx#L145-L150

This prop seems to be redundant, is there a reason why it is been used?

Comment (armenzg) [reply ui/perfherder/tests/TestsTableControls.jsx:82]: @ionutgoldan @airimovici do you have an answer for this?

Comment (camd) [reply ui/perfherder/tests/TestsTableControls.jsx:82]: @ionutgoldan @airimovici : re-ping on this question to you.  :)

**Comment by camd** [comment]
@bolah2009 : I am sorry I have taken so long to get back to you on this.  Your work here looks great to me.  But there have been a few changes since your last commit and there are some conflicts.  Would you mind rebasing this on the latest `master` and bringing it up to date with the current code base?  After we get an answer from @ionutgoldan on the perfherder parts, I will do a complete retest of the code to be sure it is solid and then we can merge this.

Thank you so much for your work and your patience!

-Camd

File: ui/infra-compare/InfraCompareTableView.jsx [hunk 3]
```diff
@@ -243,11 +243,23 @@ InfraCompareTableView.propTypes = {
     originalRevision: PropTypes.string,
     selectedTimeRange: PropTypes.string,
     updateParams: PropTypes.func.isRequired,
+    pageTitle: PropTypes.string,
   }),
   getDisplayResults: PropTypes.func.isRequired,
   getQueryParams: PropTypes.func.isRequired,
+  compareData: PropTypes.shape({
```
Comment (camd) [inline ui/infra-compare/InfraCompareTableView.jsx:250]: When I navigated to this page, I got:

    Warning: Failed prop type: Invalid prop `compareData` of type `Array` supplied to `CompareTableView`, expected instance of `Map`.
        in CompareTableView (at CompareView.jsx:231)

**Comment by bolah2009** [comment]
Hi, @camd I have updated it and resolve the conflicts. The rebase introduced new linter (prop-types) issues but it's fixed now. Thank you.

File: ui/infra-compare/InfraCompareTableView.jsx (hunk 3 above)
Comment (bolah2009) [reply ui/infra-compare/InfraCompareTableView.jsx:250]: Okay, I will check again. Thanks

Comment (bolah2009) [reply ui/infra-compare/InfraCompareTableView.jsx:250]: The `compareData` seems to be a Map, should I use `Array` instead? 🤔 

image

@camd

Comment (bolah2009) [reply ui/infra-compare/InfraCompareTableView.jsx:250]: I will use `compareData: PropTypes.instanceOf(Map)` like i did here 

https://github.com/mozilla/treeherder/blob/5cee05112a9b86afa65f9b8999ec489d00e6a205/ui/perfherder/compare/CompareTableView.jsx#L369

File: ui/infra-compare/InfraCompareTableView.jsx [hunk 4]
```diff
@@ -247,9 +247,10 @@ InfraCompareTableView.propTypes = {
   }),
   getDisplayResults: PropTypes.func.isRequired,
   getQueryParams: PropTypes.func.isRequired,
-  compareData: PropTypes.shape({
-    size: PropTypes.number,
-  }),
+  compareData: PropTypes.oneOfType([
+    PropTypes.instanceOf(Map),
+    PropTypes.array,
+  ]),
```
Comment (bolah2009) [inline ui/infra-compare/InfraCompareTableView.jsx:253]: I was still having issues using `compareData: PropTypes.instanceOf(Map)` as explained below so i went with `oneOfType`.

When I use `compareData: PropTypes.array`, I have this warning:

image


And when I use `compareData: PropTypes.instanceOf(Map)`, I have this warning:

image

**Comment by bolah2009** [comment]
Hi @camd 🙋🏽‍♂️
Made the requested changes, I have checked other parts of the UI I have access to, please let me know if am missing other parts too.

File: ui/infra-compare/InfraCompareTableView.jsx (hunk 3 above)
Comment (camd) [reply ui/infra-compare/InfraCompareTableView.jsx]: Sounds good.  I don't see the warnings anymore.  :)

File: ui/infra-compare/InfraCompareTableView.jsx (hunk 4 above)
Comment (camd) [reply ui/infra-compare/InfraCompareTableView.jsx:253]: Yeah, using `oneOfType` sounds like the right solution.

**Comment by bolah2009** [comment]
Thanks @camd

**Comment by camd** [comment]
@bolah2009 : I'm sorry to ask you to rebase this again, I'll ping @ionutgoldan on this issue after you rebase again and hopefully we can get this merged so it will stop bitrotting...

**Comment by bolah2009** [comment]
Hi @camd, I have rebased but there are new linter issues with the rebased changes. I am working on the new linter errors, I will ping you when I am through with the changes. Thanks!

I have one question, I am trying to add prop validation for these two props but I can't get the data to show them, I can only see the parent component, `InfraCompareTable`

https://github.com/mozilla/treeherder/blob/d6de0fbf79fcf49790a46736269d223bff96807e/ui/infra-compare/InfraCompareTableRow.jsx#L20-L21

**Comment by armenzg** [comment]
If I may, perhaps disabling linting at the file level for files that are related to Perfherder.
That way you don't wait on Ionut and he can follow up if he wishes to.

Just an idea. Whatever you folks decide.

**Comment by camd** [comment]
> If I may, perhaps disabling linting at the file level for files that are related to Perfherder.
> That way you don't wait on Ionut and he can follow up if he wishes to.
> 
> Just an idea. Whatever you folks decide.

Yeah, this is a great idea.  Let's do that.  Thanks, Armen, for being the voice of reason.  :)

**Comment by bolah2009** [comment]
Hi, @camd am sorry it took so long to fix the new errors after rebasing. I just updated it again and fix the newer ones too. Thanks!

File: ui/perfherder/tests/TestsTableControls.jsx (hunk 2 above)
Comment (ionutgoldan) [reply ui/perfherder/tests/TestsTableControls.jsx:82]: I really don't know how that property ended up there. @bolah2009 feel free to remove it, as it's confusing.

**Review by ionutgoldan — COMMENTED** [review]
Left a message at the top of the PR.

File: ui/perfherder/tests/TestsTableControls.jsx (hunk 2 above)
Comment (bolah2009) [reply ui/perfherder/tests/TestsTableControls.jsx:82]: Okay, thanks @ionutgoldan

**Comment by camd** [comment]
I won't get time to do any more work on this before I leave the company next Thursday.  Since this is so huge, it may be hard and risky to do this all at once.  I might suggest fixing just a sub-section of these violations, then enabling the rule and make an exclusion for all the rest of the files.  Then you can just enable and fix each file or set of files rather than having one huge PR.  It would also reduce the risk, since you can test just the affected area.

**Comment by bolah2009** [comment]
Okay @camd. Thanks. Should I close this PR and open another one or I should continue with this PR?

**Comment by camd** [comment]
> Okay @camd. Thanks. Should I close this PR and open another one or I should continue with this PR?

Hi @bolah2009 : Yes, I think that would be a good idea.  You can create a new branch for the smaller list of changes based off this one, then submit a new set of PRs for each set.

Thank you so much for your diligence with this.  I should have thought to do this in the first place.

**Comment by bolah2009** [comment]
Thanks @camd!
